include/unifex/async_auto_reset_event.hpp

```diff
@@ -0,0 +1,150 @@
+    , event_(startReady) {}
+
+  /**
+   * Puts the event in the set state if it's not in the done state.
+   *
+   * Signals any waiting *Senders* if the event was unset.
+   */
+  void set() noexcept;
+
+  /**
+   * Puts the event in the done state.
+   *
+   * Signals any waiting *Senders* if the event was unset.
+   */
+  void set_done() noexcept;
+
+  struct stream_view;
+  /**
+   * Retrieves a *Stream*-shaped view of the event.
+   */
+  stream_view stream() noexcept;
+
+private:
+  enum class state { UNSET, SET, DONE };
+  /**
+   * Alternatives to using `mutex`:
+   *   1. "2 stage commit" with `atomic<state>`, requires re-try
+   *   2. drop `set_done()` support, undesirable
+   * @see discussion on D34156868 for reasoning.
+   */
+  std::mutex mutex_;
+  state state_{state::UNSET};
+  unifex::async_manual_reset_event event_;
+
+  /**
+   * Tries to move the event from set to unset.
+   *
+   * Returns true if the event has been unset, or false if it's been
+   * cancelled.
+   */
+  bool try_reset() noexcept;
+};
+
+struct async_auto_reset_event::stream_view final {
+  explicit stream_view(async_auto_reset_event* evt) noexcept : evt_(evt) {
+    UNIFEX_ASSERT(evt_ != nullptr);
+  }
+
+  /**
+   * Returns a *Sender* that completes with set_value when the event is set.
+   */
+  auto next() noexcept {
+    return unifex::let_value_with_stop_token([evt = evt_](
+                                                 auto stopToken) noexcept {
+      return unifex::let_value_with(
+          [stopToken, evt]() noexcept {
+            // unifex::async_manual_reset_event::async_wait() returns an
+            // unstoppable *Sender* so, to support prompt cancellation
+            // of the *Sender* we're composing here, we register a stop
+            // callback with our *Receiver* that responds to stop
+            // requests by transitioning the auto-reset event to the
+            // done state. When the auto-reset event transitions to the
+            // done state, anyone waiting on our next-sender will wake
+            // up and complete with set_done.
+            //
+            // This strategy matches the *Stream* contract because
+            // cancelling the result of next(stream) is expected to
+            // cancel the whole stream.  The streams returned from
+            // event.stream() are light weight handles to the overall
+            // event and are indistinguishable from each other so
+            // cancelling one such stream by cancelling its next-sender
+            // is interpreted as cancelling the whole event.
+            auto stopCallback = [evt]() noexcept {
+              evt->set_done();
+            };
+
+            using stop_token_t = unifex::remove_cvref_t<decltype(stopToken)>;
+            using stop_callback_t =
+                typename stop_token_t::template callback_type<
+                    decltype(stopCallback)>;
+
+            return stop_callback_t{stopToken, stopCallback};
```

Comment:
I think we've learned this is a non-functional way to do stop callbacks, right?  There's probably a lifetime issue with synchronous cancellation here.

Reply:
We looked at this in detail together the other day and decided it's fine:
 - when cancelling on the same _Scheduler_ that's waiting on `async_wait()`:
    - ignoring inline _Schedulers_, the stop callback never synchronously wakes the event because `evt->set_done()` always schedules the wakeup
    - accounting for inline _Schedulers_: stop hitting yourself
 - when cancelling from a different _Scheduler_ than the one that's waiting on `async_wait()`:
    - there's technically a race between completing the stop callback and completing the `async_wait()` _but_ the `let_value_with` operation state will not complete the downstream _Receiver_ until the stop callback has been destroyed, which will synchronize with the completion of the callback; given this scenario is "cancel and complete on different _Schedulers_", the synchronization will block rather than no-op